### ISSUE TITLE
Forbid union and tag enum fields appearing in different order

### ISSUE THREAD

**Issue by Vexu** [ziglang/zig#12556]
Continued from #12543.

This used to cause bugs in stage2 (and still might) and may also just be confusing.
```zig
const E = enum {
    a,
    b,
};
const U = union(E) {
    b: u32,
    a: void,
};
```

A simple solution is to just add an error like `union field 'a'  does appear in same order as its enum tag`.

**Comment by andrewrk** [comment]
I'm accepting this because it is fundamentally simpler, i.e. according to information theory. That makes this the null hypothesis. If the other way is desired, then there needs to be a proposal for it to be that way.

**Comment by andrewrk** [comment]
Landed in ada0010471163a3accca8976185fbb6bb59c914f.
